bugtool and debuginfo

aalemayhu opened the pull request:
Please note that this is a work in progress and several things listed below are missing. I wanted to get this out earlier instead of waiting. So feedback is appreciated.

For now the cilium specific info is in `/debuginfo` which can also be retrieved by running `cilium debuginfo`. The new binary `cilium-bugtool` adds the other parts. Running the bugtool creates a ZIP file with all the files and command line output. You can optionally serve the files over HTTP with `cilium-bugtool --serve --port 80`.

aalemayhu committed:
api: new endpoint for extracting debug information from agent

- `GET /debuginfo` returns the `DebugInfo` structure which contains
  - Cilium version
  - Kernel version
  - Cilium status
  - Endpoint list
  - Service list
  - Policy
  - Cilium process memory map
  - Cilium node monitor process memory map
  - Environment variables

This is will have two users. The first one is new subcommand `cilium
debuginfo` and the second is `cilium-bugtool`.

Related: #1942 (bugtool - tool to capture node/system information)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
daemon: populate debuginfo structure

Use all of the refactored functions to retrieve status, endpoints,
rules, etc.

Related: #1942 (bugtool - tool to capture node/system information)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
cilium: add subcommand for reading debuginfo

This is the first consumer of `GET /debuginfo`.

    $ cilium debuginfo --help

    Request available debugging information from agent

    Usage:
      cilium debuginfo [flags]

    Flags:
      -f, --file string   Redirect output to file

    Global Flags:
          --config string   config file (default is $HOME/.cilium.yaml)
      -D, --debug           Enable debug messages
      -H, --host string     URI to server-side API

Also regenerate the command reference.

Related: #1942 (bugtool - tool to capture node/system information)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
version: add a structure to read out some of the version fields

This will be used by `cilium-bugtool` and `cilium debuginfo`.  While
it's added to simplify the version printing it can be used to add a JSON
format, but was not needed for now so left the FIXME alone.

The test is there to detect changes to the format in case someone starts
depending on this order.

Related: #1942 (bugtool - tool to capture node/system information)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
status: move retrieving status into a separate function

This makes it easier for other call sites on the agent side interested
in the status structure to read it without performing a HTTP request.
Will be used in `cilium debuginfo`.

Related: #1942 (bugtool - tool to capture node/system information)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
cilium-bugtool: new tool for extracting debug info from environment

    $ cilium-bugtool --help

    cilium-bugtool - capture system and node information for debugging a Cilium node

    Usage:
      bugtool [flags]

    Flags:
      -H, --host string            URI to server-side API
          --k8s-label string       Kubernetes label for Cilium pod (default "k8s-app=cilium")
          --k8s-namespace string   Kubernetes namespace for Cilium pod (default "kube-system")
      -p, --port int               Port to use for the HTTP server, (default 4444) (default 4444)
          --serve                  Start HTTP server to serve static files
      -t, --tmp string             Path to store extracted files (default "/tmp/")

Closes: #1942 (bugtool - tool to capture node/system information)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
endpoint: refactor retrieving and print endpoint list

Makes it easier to use from `GET /debuginfo`.

No functional change intended.

Related: #1942 (bugtool - tool to capture node/system information)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
service: refactor so getting service list outside is easier

used in `GET /debuginfo`.

No functional change intended.

Related: #1942 (bugtool - tool to capture node/system information)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
contrib: add cilium-bugtool to docker cp-dirs

We want this to be available from the container.

Related: #1942 (bugtool - tool to capture node/system information)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
monitor: expose the PID

will be used in `GET /debuginfo` to read the memory map.

Related: #1942 (bugtool - tool to capture node/system information)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

tgraf commented on daemon/debuginfo.go:
```diff
@@ -0,0 +1,91 @@
+	return string(m)
+}
+
+func getRulesList(d *Daemon) *models.Policy {
```
Let's move this function into `pkg/policy/repository.go`. That will make it easier to make the repository mutex private at a later point.

tgraf commented on cilium/cmd/debuginfo.go:
```diff
@@ -0,0 +1,90 @@
+	if resp, err := client.Daemon.GetDebuginfo(nil); err != nil {
+		fmt.Fprintf(os.Stderr, "%s\n", pkg.Hint(err))
+		os.Exit(1)
+	} else {
```
You can avoid this else branch as you will quit the program above.

aalemayhu replied:
Sure, was copy pasta.

tgraf commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,297 @@
+	}
+	// We are using ZIP format because it's widely supported.
+	output := filepath.Join(tmpDirPrefix, dbgDir+".zip")
+	zip.ArchiveFile(dbgDir, output, progress)
```
I'm undecided here. On one side I like not relying on additional external tools. OTOH, this adds more external dependencies to our already large list. Maybe just invoke `tar` for this?

aalemayhu replied:
I was using `tar` earlier but had some issue with directories where it would just create empty files. Can look into it again and revert.

tgraf commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,297 @@
+func runTool() {
+	var dbgDir string
+	// Prevent collision with other directories
+	d, err := ioutil.TempDir(dumpPath, "bugtool-")
```
prefix this with `cilium-bugtool-`

tgraf commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,297 @@
+	if err != nil {
+		log.WithError(err).Warn("failed to create debug directory")
+	}
+	dbgDir = d
```
Why are you assigning first to `d` and then `dbgDir`?

aalemayhu replied:
I had a fallback directory to tmp but it was removed.

tgraf commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,297 @@
+	sources := []string{
+		"hostname", "ip a", "ip r", "ip link", "iptables-save", "uname -a",
+		"uptime", "dmesg", "ethtool", "cilium debuginfo", "bpftool map show",
+		"bpftool prog show", "iptables -S", "ip6tables -S", "ptables -L -v",
```
typo: `ptables`

tgraf commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,297 @@
+		"hostname", "ip a", "ip r", "ip link", "iptables-save", "uname -a",
+		"uptime", "dmesg", "ethtool", "cilium debuginfo", "bpftool map show",
+		"bpftool prog show", "iptables -S", "ip6tables -S", "ptables -L -v",
+		"cilium config", "ps", "docker version", "kubectl version", "env",
```
Add `cilium bpf` programs here as well?

tgraf commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,297 @@
+		access logs
+	*/
+
+	copyResolveConfig(dbgDir)
```
Can we use `copyFile`?

aalemayhu replied:
Yes

tgraf commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,297 @@
+
+	if serve {
+		serveStaticFiles(dbgDir)
+		return
```
Is it intentional to keep the tmp directory around after serving?

aalemayhu replied:
Yes was mostly to check files were in there, but removed.

tgraf reviewed (dismissed):
Fantastic PR! Very clean and great commit messages. Some feedback as comments.

aalemayhu committed:
policy: add function for retrieving the rules list

Will be used in `GET /debuginfo`.

No functional change intended.

Related: #1942 (bugtool - tool to capture node/system information)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aanm commented on pkg/version/version.go:
```diff
@@ -17,13 +17,45 @@ package version
+	Revision string
+	// GoRuntimeVersion is the Go version used to run Cilium
+	GoRuntimeVersion string
+	// Arch is the archtectirure where Cilium was compiled
```
`s/archtectirure/architecture/`

aanm commented on cilium/cmd/service_list.go:
```diff
@@ -44,7 +45,10 @@ func listServices(cmd *cobra.Command, args []string) {
 	if err != nil {
 		Fatalf("Cannot get services list: %s", err)
 	}
+	printServicList(list)
```
`s/printServicList/printServiceList`

aanm commented on daemon/loadbalancer.go:
```diff
@@ -343,17 +343,20 @@ func NewGetServiceHandler(d *Daemon) GetServiceHandler {
+	return NewGetServiceOK().WithPayload(list)
+}
 
+func getServiceList(d *Daemon) []*models.Service {
```
Why not use Daemon as the receiver?

aalemayhu replied:
Please elaborate. I am not sure I am following. I wrapped the getServiceList lines so that there would be no need for extra HTTP calls.

aalemayhu replied:
Do you mean move the function so it is called from the daemon?
```go
+(d *Daemon) func getServiceList() []*models.Service {
```

aanm replied:
yes

aalemayhu replied:
Looks better, thanks.

aanm commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,297 @@
+// CopyFile copies a source file to a destination.
+//
+// It follows symbolic links and retains modes.
+func CopyFile(source string, dest string) error {
```
Usually dest is the 1st argument and source is 2nd, for example in `io.Copy`

aanm commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,297 @@
+	}
+}
+
+func serveStaticFiles(debugDirectory string) {
```
What will be this http server for?

aalemayhu replied:
I wanted to have this option so we maybe could expand on it for HTML rendering. This will just serve the files and IMO can make digesting the content easier if you can open all the files you are interested in f. ex. a browser. Would prefer to keep it in. It's optional so should not interfere with people who just want the archive of files.

aanm replied:
But still you'll be opening a port on the users' machine, disable this by default.

aalemayhu replied:
Not sure what you mean? It is already disabled by default. You need to pass it the `--serve` option for it to listen on that port.

aanm replied:
Disregard my comment then ;-)

aanm commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,297 @@
+
+	copyResolveConfig(dbgDir)
+	copyKernelConfig(dbgDir)
+	createArchive(dbgDir, dumpPath)
```
How big is the file? I'm just trying to understand if it makes easier to paste everything into a single `.txt` file or multiple ones in a `.zip`. If it's a text file it's easier for users to just copy-paste it to the bug report on github.

tgraf replied:
These files will be very big. We may want to have a summary which includes the essential information. We can have the bugtool concat that together based on the individual files.

aalemayhu replied:
So the file is 110M on my workstation with no endpoints running, but this will be very host specific considering we are getting a lot of different command output. 

IMO the bugtool should gather as much information as it can. Then the user should pick what they are interested in. I don't think summarizing is a good idea. It's hard to know what exactly someone using the tool wants. If it gets too slow we can add filtering and / or use several archives as @aanm suggests.

aalemayhu replied:
On my workstation the biggest files are `lsof` and `journalctl -u docker`. 
Also tried on a smaller docker deployment and there the archive is 5.2M and the biggest file is `cat /proc/kallsyms`.

aanm replied:
We probably don't need docker's logs. IMO `lsof` should be optional. We will probably get 95% of the information we want from `iptables-save`, `cilium logs`, `cilium endpoint list`, `uname -a`, `kubectl get version` and a couple more that I don't remember this moment.

aalemayhu replied:
Dropping `lsof` then. Want to keep the docker logs just in case. Not sure yet of a good way to make these optional.

aanm replied:
@scanf why the docker logs?

aalemayhu replied:
I was thinking it might capture useful information on environments where the cilium deployment is not called cilium but something else. Can drop it.

aalemayhu committed:
Documentation: show usage examples of bugtool and debuginfo

Related: #1942 (bugtool - tool to capture node/system information)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
tests: copy the debuginfo and cilium-bugtool output for CI artifact

Related: #1942 (bugtool - tool to capture node/system information)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

raybejjani commented on Documentation/troubleshooting.rst:
```diff
@@ -134,6 +134,75 @@ running agent. Debugging of an individual endpoint can be enabled by running
+
+The ``cilium-bugtool`` let's you capture possibly useful informations about
+your environment for debugging. The tool is meant to be used for debugging a
+singel Cilium agent node. It collects command outputs and files from several
```
singel

aalemayhu replied:
Not sure what you mean, but this is just a draft which will be revisited when I am done with the review and remaining items for dropping the RFC tag.

raybejjani replied:
sorry. I was unhelpfully pointing out the typo. I keep seeing review request emails from github and forget to check the wip tag.

raybejjani commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,237 @@
+	// Wait for the commands to complete
+	for i := 0; i < len(sources); i++ {
+		<-execs
+	}
```
Is the intention to force the commands to run one by one? If so, why not just spawn one goroutine that does the `for _, cmd := range sources` loop but executes the commands directly?
If the goal is to later allow more than one to run at once, this is a good way to do it (although in that case, I would recommend `make(chan bool, 1)` to make that more clear).

aalemayhu replied:
Will change it to the later.

raybejjani replied:
Oh, in fact, I misread this. Because we write to the channel after we run the command, all we were/are doing was making the main routine wait for all the spawned goroutines to complete. What you have will do this, or you can use a WaitGroup (https://golang.org/pkg/sync/#WaitGroup) that does pretty must the same thing.

If the goal was to have only 2 (or whatever) commands run at once it's a little more complicated but:
```golang
	numRoutinesAtOnce := 2
	semaphore := make(chan bool, numRoutinesAtOnce)
	for i := 0; i < numRoutinesAtOnce; i++ {
		// This will not block because the channel is buffered and we can write to it numRoutinesAtOnce before the write blocks
		semaphore <- true
	}

	for _, cmd := range commands {
		wg := syncWaitGroup()
		// tell the waitgroup it needs to track another goroutine
		wg.Add(1)

		// start a subroutine to run our command
		go func() {
			// once we exit this goroutine completely, signal the original that we are done
			defer wg.Done()

			// This will wait until an entry in this channel is available to read. We started with numRoutinesAtOnce in there (from above)
			<-semaphore
			// When we are done we return the thing we took from the semaphore, so another goroutine can get it
			defer func() { semaphore <- true }()
			cmd()
		}()
	}
	// wait for all the spawned goroutines to finish up.
	wg.Wait()
```

raybejjani commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,237 @@
+	fmt.Printf("exec: %s %s\n", cmd, args)
+	output, err := exec.Command(cmd, args...).CombinedOutput()
+	if err != nil {
+		return string(output), nil
```
Should this return err? If not, we could skip the if completely.

aalemayhu replied:
It should I just missed it while moving around lines. Thanks.

raybejjani commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,237 @@
+	// Clean up the filename
+	name := strings.Replace(prompt, "/", " ", -1)
+	name = strings.Replace(name, " ", "-", -1)
+	f, _ := os.Create(filepath.Join(cmdDir, name+".md"))
```
It might be better to exit on error, since f is undefined in that case (or nil) and that might cause weird panics later when we try to use f.

aalemayhu replied:
Yeah, we should return and log it.

raybejjani commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,237 @@
+		os.Remove(f.Name())
+		return
+	}
+	f.Close()
```
This might have to be defered earlier since we return early just above. Or we can ditch the return to ensure we get here.

raybejjani commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,237 @@
+		serveStaticFiles(dbgDir)
+	}
+
+	if err := os.RemoveAll(dbgDir); err != nil {
```
I'm not sure, but I think ListenAndServe never returns (in serveStaticFiles) and if that's the case then the only way to exit would be with a ctrl-C or something which then would mean we never run this cleanup step. Not that big a deal since these are in temp directories.
I think there are ListenAndServe wrappers or versions that allow exit. If not, we can run that in a goroutine, and wait here for some other trigger (possibly a os.Signal)

aalemayhu replied:
Good catch. Did not notice it would skip it. For the temporary files as you point it's not a big issue but we should always print the disclaimer. Changing it to defer by default but also support `SIGINT`.

raybejjani reviewed (commented):
I think this will work out well :)

tgraf commented on daemon/accesslog.go:
```diff
@@ -0,0 +1,53 @@
+	path := viper.GetString(accesLogKey)
+	dr.Path = path
+
+	content, err := ioutil.ReadFile(path)
```
I don't think that this is a good idea.
1) I'm not sure it makes sense to expose the full access log via the API at this point.
2) If we want to add this, it should be added as a streaming API where only the access log stream is returned.

I would return the path via `DaemonConfigurationResponse` as part of the `GET /config` API

aalemayhu replied:
Dropping this then. I can write follow up issues later.

raybejjani commented on bugtool/cmd/helper.go:
```diff
@@ -0,0 +1,144 @@
+		return
+	}
+
+	_ = filepath.Walk(dbgDir, func(path string, info os.FileInfo, err error) error {
```
If we won't handle the error we could ditch the `_ = ` bit. We might want to, though, since it would error out if the directory doesn't exist or we have bad permissions etc.

aalemayhu replied:
Will change it so we handle the error.

raybejjani commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,354 @@
+func init() {
+	RootCmd.Flags().BoolVar(&serve, "serve", false, "Start HTTP server to serve static files")
+	RootCmd.Flags().IntVarP(&port, "port", "p",
+		4444, "Port to use for the HTTP server, (default 4444)")
```
I think this was split to meet a 80 char limit but line 60 doesn't. I'd vote to have it run long and be on one line (but go has something like a 120 char limit, if any).

aalemayhu replied:
Can't remember why this was done. Not intentional.

raybejjani commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,354 @@
+	output, err := execCommand("kubectl", "-n", namespace, "get", "pods", "-l", label)
+	ciliumPods := []string{}
+	if err != nil {
+		return ciliumPods, err
```
you could return the literal directly here (`return []string{}, err`) since it isn't used otherwise. Normally if a function returns an error it isn't expected that the other return variables are valid so you could also return a nil (`return nil, err`). It just so happens that nil is an empty slice so most operations (like len, range and cap) will return reasonable values, so it isn't as bad as a null would be in C.

aalemayhu replied:
Sounds good.

raybejjani commented on cilium/cmd/status.go:
```diff
@@ -69,29 +56,51 @@ func statusDaemon(cmd *cobra.Command, args []string) {
-			for _, ipv6 := range sr.IPAM.IPV6 {
-				fmt.Fprintf(w, " %s\n", ipv6)
-			}
+func printStatus(w *tabwriter.Writer, sr *models.StatusResponse) {
```
+1 to making this a function

aalemayhu replied:
Will drop this function since there is [FormatStatusResponse](https://godoc.org/github.com/cilium/cilium/pkg/client#FormatStatusResponse) now.

aalemayhu committed:
cilium: add bugtool as subcommand

Also regenerate command reference.

Requested-by: André Martins <andre@cilium.io>
Related: #1942 (bugtool - tool to capture node/system information)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

ianvernon commented on monitor/launch/launcher.go:
```diff
@@ -37,6 +37,11 @@ type NodeMonitor struct {
 	state   *models.MonitorStatus
 }
 
+// GetPid returns the node monitors pid.
```
Nit: node monitor's

ianvernon commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,417 @@
+
+Thanks.
+
+[0]: http://docs.cilium.io/en/stable/troubleshooting/#single-node-bugtool
```
If this changes over time and someone is using a new version, will that be an issue? Is there a way we can make the version of the documentation dependent upon the version of the agent currently running?

aalemayhu replied:
Sorry, I don't have a good answer for this yet. Will think about it some more, but if you have any ideas please share.

ianvernon commented on bugtool/cmd/root.go:
```diff
@@ -0,0 +1,417 @@
+	}
+}
+
+// split takes a commmand prompt and returns a the command and arguments seperately
```
Nit: returns the

ianvernon commented on Documentation/troubleshooting.rst:
```diff
@@ -134,6 +134,104 @@ running agent. Debugging of an individual endpoint can be enabled by running
+Debugging information
+=====================
+
+``cilium debuginfo`` can print  useful output from the Cilium API. The output
```
remove extra space between 'print  useful'

ianvernon commented on Documentation/troubleshooting.rst:
```diff
@@ -134,6 +134,104 @@ running agent. Debugging of an individual endpoint can be enabled by running
+=====================
+
+``cilium debuginfo`` can print  useful output from the Cilium API. The output
+format is Markdown format so this can be used when reporting a bug on the
```
Change to "The output is in Markdown format"

ianvernon commented on Documentation/troubleshooting.rst:
```diff
@@ -134,6 +134,104 @@ running agent. Debugging of an individual endpoint can be enabled by running
+environment for debugging. The tool is meant to be used for debugging a single
+Cilium agent node but in the Kubernetes case if you have multiple Cilium pods
+it can retrieve debugging information from all of them. The tool works by
+collecting command outputs and files from several places and creates archive.
```
... works by archiving a collection of command outputs and files from several places.

ianvernon commented on Documentation/troubleshooting.rst:
```diff
@@ -134,6 +134,104 @@ running agent. Debugging of an individual endpoint can be enabled by running
+Single Node Bugtool
+===================
+
+The ``cilium-bugtool`` captures possibly useful informations about your
```
possibly --> potentially

ianvernon commented on Documentation/troubleshooting.rst:
```diff
@@ -134,6 +134,104 @@ running agent. Debugging of an individual endpoint can be enabled by running
+
+  cilium-bugtool
+
+Running it with no option as shown above it will try to copy various files and
```
When running it with no options as shown above, it will...

ianvernon commented on tests/helpers.bash:
```diff
@@ -623,6 +625,9 @@ function dump_cli_output_k8s {
   kubectl exec -n ${NAMESPACE} ${POD} -- cilium status > ${DIR}/${POD}_status.txt
+  kubectl exec -n ${NAMESPACE} ${POD} -- cilium debuginfo > ${DIR}/${POD}_debuginfo.txt
+  local DEBUGTOOL_ARCHIVE=`kubectl exec -n ${NAMESPACE} ${POD} -- cilium-bugtool | grep ARCHIVE | awk '{ print $3}'`
+  kubectl cp ${NAMESPACE}/${POD}:${DEBUGTOOL_ARCHIVE} ${DIR}/${POD}_bugtool.tar
```
Could we potentially just get rid of lines 627 and above and just rely upon the archive created by the debugtool?

aalemayhu replied:
The archive should have the equivalent output + more. If you want I can drop the `debuginfo` call and remove the other text files that are covered.

aalemayhu replied:
Could also be done as follow up, will add it to #2165 with a question mark.

ianvernon reviewed (changes requested):
Mainly small cosmetic changes. Overall, this looks great! Clearly commented, documented, and organized well. Well done!
